### PR TITLE
nad pkg: added MasterBouldPlugin struct

### PR DESCRIPTION
**Pull request by gkopels**

### CODE REVIEW

File: pkg/nad/masterplugin.go [hunk 1]
```diff
@@ -393,3 +393,65 @@ func (plugin *MasterIPVlanPlugin) GetMasterPluginConfig() (*MasterPlugin, error)
 
 	return plugin.masterPlugin, nil
 }
+
+// MasterBoundPlugin provides struct for MasterPlugin to create a NeNetworkAttachmentDefinition Master plugin for bonded interfaces.
```
Comment (evgenLevin) [inline pkg/nad/masterplugin.go]: Typo? MasterBondPlugin

File: pkg/nad/masterplugin.go [hunk 2]
```diff
@@ -393,3 +393,65 @@ func (plugin *MasterIPVlanPlugin) GetMasterPluginConfig() (*MasterPlugin, error)
...
+type MasterBoundPlugin struct {
+	masterPlugin *MasterPlugin
+	errorMsg     string
+}
+
+// NewMasterBoundPlugin define network-attachment-interface for bond interface.
```
Comment (evgenLevin) [inline pkg/nad/masterplugin.go:404]: typo: defines

File: pkg/nad/masterplugin.go [hunk 3]
```diff
@@ -393,3 +393,65 @@ func (plugin *MasterIPVlanPlugin) GetMasterPluginConfig() (*MasterPlugin, error)
...
+		masterPlugin: &MasterPlugin{
+			CniVersion:      "0.3.1",
+			Name:            name,
+			Type:            "bond",
+			Mode:            mode,
+			FailOverMac:     1,
+			LinkInContainer: true,
+			Miimon:          "100",
+			Links:           []Link{{Name: "net1"}, {Name: "net2"}},
+			Capabilities:    &Capability{IP: true},
```
Comment (evgenLevin) [inline pkg/nad/masterplugin.go:419]: Please proivede next values as the func parameters:

- FaileOverMac
- Miimon
- Links

File: pkg/nad/masterplugin.go [hunk 4]
```diff
@@ -393,3 +393,65 @@ func (plugin *MasterIPVlanPlugin) GetMasterPluginConfig() (*MasterPlugin, error)
...
+		},
+	}
+
+	if builder.masterPlugin.Name == "" {
+		glog.V(100).Infof("error NewMasterBoundPlugin name can not be empty")
+
+		builder.errorMsg = "MasterMacVlanPlugin name is empty"
+	}
+
+	if builder.masterPlugin.Mode == "" {
```
Comment (evgenLevin) [inline pkg/nad/masterplugin.go:429]: I think you need create a list of supported mode and validate that this mode is in this list

File: pkg/nad/masterplugin.go [hunk 5]
```diff
@@ -393,3 +393,66 @@ func (plugin *MasterIPVlanPlugin) GetMasterPluginConfig() (*MasterPlugin, error)
...
+
+		builder.errorMsg = "NewMasterBoundPlugin Mode is empty"
+	}
+
+	return &builder
+}
+
+// WithMTU defines nad with MTU.
+func (plugin *MasterBoundPlugin) WithMTU(mtu string) *MasterBoundPlugin {
+	if mtu == "" {
```
Comment (evgenLevin) [inline pkg/nad/masterplugin.go:440]: Validate that the builder doesnt have errors
	if valid, _ := builder.validate(); !valid {
		return builder
	}

Comment (evgenLevin) [inline pkg/nad/masterplugin.go:440]: You need to validate that mtu in the supported MTU range 1-9192. For that you need to change MTU parameter type to int and chack that this value in range:
func (plugin *MasterBoundPlugin) WithMTU(mtu int) *MasterBoundPlugin {
...
	if 1 > mtu || mtu > 9192 {
		builder.errorMsg = fmt.Sprintf("invalid mtu size %d allowed mtu should be in range 1...9192", mtu)
	}

File: pkg/nad/masterplugin.go [hunk 6]
```diff
@@ -393,3 +393,66 @@ func (plugin *MasterIPVlanPlugin) GetMasterPluginConfig() (*MasterPlugin, error)
...
+		glog.V(100).Infof("error adding empty mtu to MasterBoundPlugin")
+
+		plugin.errorMsg = "error adding empty mtu to MasterBoundPlugin"
+	}
+
```
Comment (evgenLevin) [inline pkg/nad/masterplugin.go:445]: Please chacke there is no errors:
	if builder.errorMsg != "" {
		return builder
	}

**Review by evgenLevin — COMMENTED** [review]
Several comments

**Review by kononovn — CHANGES_REQUESTED** [review]
Hello @gkopels. Thanks for you contribution. There is no "cnf network" under this project. Please change PR title to "nad pkg: added MasterBouldPlugin struct". Please update commit message as well. Thanks

File: pkg/nad/masterplugin.go (hunk 1 above)
Comment (gkopels) [reply pkg/nad/masterplugin.go]: Changed all references to MasterBondPlugin

File: pkg/nad/masterplugin.go (hunk 2 above)
Comment (gkopels) [reply pkg/nad/masterplugin.go]: Changed

File: pkg/nad/masterplugin.go (hunk 3 above)
Comment (gkopels) [reply pkg/nad/masterplugin.go]: Created separate functions for each value

File: pkg/nad/masterplugin.go (hunk 4 above)
Comment (gkopels) [reply pkg/nad/masterplugin.go]: added
		"balance-rr":    true,
		"active-backup": true,
		"balance-xor":   true,

File: pkg/nad/masterplugin.go (hunk 5 above)
Comment (gkopels) [reply pkg/nad/masterplugin.go]: I want to talk to you about this

Comment (gkopels) [reply pkg/nad/masterplugin.go]: Changed

File: pkg/nad/masterplugin.go (hunk 6 above)
Comment (gkopels) [reply pkg/nad/masterplugin.go:510]: Made a change. Have a look.

File: pkg/nad/masterplugin.go [hunk 7]
```diff
@@ -393,3 +393,161 @@ func (plugin *MasterIPVlanPlugin) GetMasterPluginConfig() (*MasterPlugin, error)
...
 	return plugin.masterPlugin, nil
 }
+
+// MasterBondPlugin provides struct for MasterPlugin to create a NAD for bonded interfaces.
+type MasterBondPlugin struct {
+	masterPlugin *MasterPlugin
+	errorMsg     string
+}
+
+// NewMasterBondPlugin define network-attachment-interface for bond interface.
```
Comment (evgenLevin) [inline pkg/nad/masterplugin.go:403]: wrong  comment. Should be something like:
// NewMasterBondPlugin creates new instance of MasterBondPlugin

File: pkg/nad/masterplugin.go [hunk 8]
```diff
@@ -393,3 +393,161 @@ func (plugin *MasterIPVlanPlugin) GetMasterPluginConfig() (*MasterPlugin, error)
...
+func NewMasterBondPlugin(name string) *MasterBondPlugin {
+	glog.V(100).Infof(
+		"Initializing new MasterBondPlugin structure %s", name)
+
+	builder := MasterBondPlugin{
+		masterPlugin: &MasterPlugin{
+			CniVersion: "0.3.1",
+			Name:       name,
+			Type:       "bond",
```
Comment (evgenLevin) [inline pkg/nad/masterplugin.go:412]: Please add a mode as a default parameter

File: pkg/nad/masterplugin.go [hunk 9]
```diff
@@ -393,3 +393,161 @@ func (plugin *MasterIPVlanPlugin) GetMasterPluginConfig() (*MasterPlugin, error)
...
+	validModes := map[string]bool{
+		"balance-rr":    true,
+		"active-backup": true,
+		"balance-xor":   true,
+	}
+
+	if mode == "" {
+		glog.V(100).Infof("error adding empty mode to MasterBondPlugin")
+
+		return nil
```
Comment (evgenLevin) [inline pkg/nad/masterplugin.go:436]: no need return here, add error to the struct
plugin.errorMsg ="Bond mode can not be empty"

File: pkg/nad/masterplugin.go [hunk 10]
```diff
@@ -393,3 +393,161 @@ func (plugin *MasterIPVlanPlugin) GetMasterPluginConfig() (*MasterPlugin, error)
...
+	}
+
+	// Check if the provided mode is valid
+	if !validModes[mode] {
+		glog.V(100).Infof("error: invalid mode type used for MasterBondPlugin bond interface")
```
Comment (evgenLevin) [inline pkg/nad/masterplugin.go:441]: Please add to the log what invalid mode was provided

File: pkg/nad/masterplugin.go [hunk 11]
```diff
@@ -393,3 +393,161 @@ func (plugin *MasterIPVlanPlugin) GetMasterPluginConfig() (*MasterPlugin, error)
...
+
+		return nil
```
Comment (evgenLevin) [inline pkg/nad/masterplugin.go:443]: no need return, add error to the struct

File: pkg/nad/masterplugin.go [hunk 12]
```diff
@@ -393,3 +393,161 @@ func (plugin *MasterIPVlanPlugin) GetMasterPluginConfig() (*MasterPlugin, error)
...
+	}
+
+	plugin.masterPlugin.Mode = mode
+
+	return plugin
+}
+
+// WithLinksInContainer defines nad with linksInContainer as true.
```
Comment (evgenLevin) [inline pkg/nad/masterplugin.go:451]: not always true, could be false as well

File: pkg/nad/masterplugin.go [hunk 13]
```diff
@@ -393,3 +393,161 @@ func (plugin *MasterIPVlanPlugin) GetMasterPluginConfig() (*MasterPlugin, error)
...
+func (plugin *MasterBondPlugin) WithLinksInContainer(linksInContainer bool) *MasterBondPlugin {
+	if !linksInContainer {
```
Comment (evgenLevin) [inline pkg/nad/masterplugin.go:453]: Why I can not configure false?

File: pkg/nad/masterplugin.go [hunk 14]
```diff
@@ -393,3 +393,161 @@ func (plugin *MasterIPVlanPlugin) GetMasterPluginConfig() (*MasterPlugin, error)
...
+		glog.V(100).Infof("error adding empty linksInContainer to MasterBondPlugin")
```
Comment (evgenLevin) [inline pkg/nad/masterplugin.go:454]: wrong log, linksInContaniner is not empty

File: pkg/nad/masterplugin.go [hunk 15]
```diff
@@ -393,3 +393,161 @@ func (plugin *MasterIPVlanPlugin) GetMasterPluginConfig() (*MasterPlugin, error)
...
+
+		plugin.errorMsg = "error adding empty linksInContainer to MasterBondPlugin"
+	}
+
+	plugin.masterPlugin.LinksInContainer = linksInContainer
+
+	return plugin
+}
+
+// WithFailOverMac defines nad with WithFailOverMac amount.
```
Comment (evgenLevin) [inline pkg/nad/masterplugin.go:464]: not amount. it could be 0 (none) , 1 (active) and 2(follow)

File: pkg/nad/masterplugin.go [hunk 16]
```diff
@@ -393,3 +393,161 @@ func (plugin *MasterIPVlanPlugin) GetMasterPluginConfig() (*MasterPlugin, error)
...
+func (plugin *MasterBondPlugin) WithFailOverMac(failOverMac int) *MasterBondPlugin {
+	if failOverMac == 0 {
```
Comment (evgenLevin) [inline pkg/nad/masterplugin.go:466]: Why? Please validate only what it's 0,1 or 2

File: pkg/nad/masterplugin.go [hunk 17]
```diff
@@ -393,3 +393,161 @@ func (plugin *MasterIPVlanPlugin) GetMasterPluginConfig() (*MasterPlugin, error)
...
+		glog.V(100).Infof("error adding empty FailOverMac to MasterBondPlugin")
```
Comment (evgenLevin) [inline pkg/nad/masterplugin.go:467]: wrong log

File: pkg/nad/masterplugin.go [hunk 18]
```diff
@@ -393,3 +393,161 @@ func (plugin *MasterIPVlanPlugin) GetMasterPluginConfig() (*MasterPlugin, error)
...
+	}
+
+	plugin.masterPlugin.FailOverMac = failOverMac
+
+	return plugin
+}
+
+// WithMiimon defines nad with WithMiimon.
+func (plugin *MasterBondPlugin) WithMiimon(miimon string) *MasterBondPlugin {
+	if miimon == "" {
```
Comment (evgenLevin) [inline pkg/nad/masterplugin.go:479]: maybe we need add validation that miimon is integer (not random srtring) and is higher than 0

File: pkg/nad/masterplugin.go [hunk 19]
```diff
@@ -393,3 +393,161 @@ func (plugin *MasterIPVlanPlugin) GetMasterPluginConfig() (*MasterPlugin, error)
...
+
+	plugin.masterPlugin.Ipam = ipam
+
+	return plugin
+}
+
+// GetMasterPluginConfig returns master plugin if error is not occur.
+func (plugin *MasterBondPlugin) GetMasterPluginConfig() (*MasterPlugin, error) {
+	if plugin.errorMsg != "" {
+		return nil, fmt.Errorf("error to build MaterPlugin config due to :%s", plugin.errorMsg)
```
Comment (evgenLevin) [inline pkg/nad/masterplugin.go:549]: MaterBondPlugin

File: pkg/nad/masterplugin.go [hunk 20]
```diff
@@ -393,3 +393,161 @@ func (plugin *MasterIPVlanPlugin) GetMasterPluginConfig() (*MasterPlugin, error)
...
+		plugin.errorMsg = "error adding empty FailOverMac to MasterBondPlugin"
+	}
+
+	plugin.masterPlugin.FailOverMac = failOverMac
+
+	return plugin
+}
+
+// WithMiimon defines nad with WithMiimon.
+func (plugin *MasterBondPlugin) WithMiimon(miimon string) *MasterBondPlugin {
```
Comment (evgenLevin) [inline pkg/nad/masterplugin.go:478]: I think better to get int miimon value and after validation convert it to string if needed

File: pkg/nad/masterplugin.go [hunk 21]
```diff
@@ -393,3 +393,161 @@ func (plugin *MasterIPVlanPlugin) GetMasterPluginConfig() (*MasterPlugin, error)
...
+	plugin.masterPlugin.Mtu = mtu
+
+	if plugin.errorMsg == "" {
+		plugin.errorMsg = "An error occurred applying the mtu value"
+	}
+
+	return plugin
+}
+
+// WithIPAM defines nad with IP type.
```
Comment (evgenLevin) [inline pkg/nad/masterplugin.go:533]: // WithIPAM defines IPAM configuration to MasterBondPlugin

File: pkg/nad/masterplugin.go [hunk 22]
```diff
@@ -393,3 +393,161 @@ func (plugin *MasterIPVlanPlugin) GetMasterPluginConfig() (*MasterPlugin, error)
...
+
+		plugin.errorMsg = "error adding empty capabilities to MasterBondPlugin"
+	}
+
+	plugin.masterPlugin.Capabilities = capabilities
+
+	return plugin
+}
+
+// WithMTU defines nad with MTU.
```
Comment (evgenLevin) [inline pkg/nad/masterplugin.go:516]: // WithMTU defines MTU configuration to MasterBondPlugin

File: pkg/nad/masterplugin.go [hunk 23]
```diff
@@ -393,3 +393,161 @@ func (plugin *MasterIPVlanPlugin) GetMasterPluginConfig() (*MasterPlugin, error)
...
+
+		plugin.errorMsg = "error adding empty links to MasterBondPlugin"
+	}
+
+	plugin.masterPlugin.Links = links
+
+	return plugin
+}
+
+// WithCapabilities defines nad with configuration for capabilities.
```
Comment (evgenLevin) [inline pkg/nad/masterplugin.go:503]: // WithCapabilities  defines Capabilities configuration to MasterBondPlugin

File: pkg/nad/masterplugin.go [hunk 24]
```diff
@@ -393,3 +393,161 @@ func (plugin *MasterIPVlanPlugin) GetMasterPluginConfig() (*MasterPlugin, error)
...
+
+		plugin.errorMsg = "error adding empty miimon to MasterBondPlugin"
+	}
+
+	plugin.masterPlugin.Miimon = miimon
+
+	return plugin
+}
+
+// WithLinks defines nad with configuration for Links.
```
Comment (evgenLevin) [inline pkg/nad/masterplugin.go:490]: WithLinks defines Links configuration to MasterBondPlugin

File: pkg/nad/masterplugin.go [hunk 25]
```diff
@@ -393,3 +393,161 @@ func (plugin *MasterIPVlanPlugin) GetMasterPluginConfig() (*MasterPlugin, error)
...
+
+		plugin.errorMsg = "error adding empty FailOverMac to MasterBondPlugin"
+	}
+
+	plugin.masterPlugin.FailOverMac = failOverMac
+
+	return plugin
+}
+
+// WithMiimon defines nad with WithMiimon.
```
Comment (evgenLevin) [inline pkg/nad/masterplugin.go:477]: WithMiimon  defines Miimon configuration to MasterBondPlugin

File: pkg/nad/masterplugin.go (hunk 15 above)
Comment (evgenLevin) [reply pkg/nad/masterplugin.go:464]: WithFailOverMac defines FailOverMac configuration to MasterBondPlugin

File: pkg/nad/masterplugin.go (hunk 12 above)
Comment (evgenLevin) [reply pkg/nad/masterplugin.go:451]: WithLinksInContainer defines linksInContainer configuration to MasterBondPlugin

File: pkg/nad/masterplugin.go (hunk 7 above)
Comment (gkopels) [reply pkg/nad/masterplugin.go:403]: changed

File: pkg/nad/masterplugin.go (hunk 11 above)
Comment (gkopels) [reply pkg/nad/masterplugin.go:443]: changed

File: pkg/nad/masterplugin.go (hunk 21 above)
Comment (gkopels) [reply pkg/nad/masterplugin.go:533]: changed

File: pkg/nad/masterplugin.go (hunk 22 above)
Comment (gkopels) [reply pkg/nad/masterplugin.go:516]: Changed

File: pkg/nad/masterplugin.go (hunk 23 above)
Comment (gkopels) [reply pkg/nad/masterplugin.go:503]: changed

File: pkg/nad/masterplugin.go (hunk 24 above)
Comment (gkopels) [reply pkg/nad/masterplugin.go:490]: changed

File: pkg/nad/masterplugin.go (hunk 25 above)
Comment (gkopels) [reply pkg/nad/masterplugin.go:477]: changed

File: pkg/nad/masterplugin.go (hunk 8 above)
Comment (gkopels) [reply pkg/nad/masterplugin.go:418]: moved

File: pkg/nad/masterplugin.go (hunk 9 above)
Comment (gkopels) [reply pkg/nad/masterplugin.go]: changed

File: pkg/nad/masterplugin.go (hunk 15 above)
Comment (gkopels) [reply pkg/nad/masterplugin.go]: changed

File: pkg/nad/masterplugin.go (hunk 19 above)
Comment (gkopels) [reply pkg/nad/masterplugin.go:520]: fixed

File: pkg/nad/masterplugin.go (hunk 20 above)
Comment (gkopels) [reply pkg/nad/masterplugin.go]: changed

File: pkg/nad/masterplugin.go [hunk 26]
```diff
@@ -393,3 +394,132 @@ func (plugin *MasterIPVlanPlugin) GetMasterPluginConfig() (*MasterPlugin, error)
...
+	}
+
+	plugin.masterPlugin.LinksInContainer = linksInContainer
+
+	return plugin
+}
+
+// WithFailOverMac defines FailOverMac configuration to MasterBondPlugin.
+func (plugin *MasterBondPlugin) WithFailOverMac(failOverMac int) *MasterBondPlugin {
+	if failOverMac > 2 {
```
Comment (evgenLevin) [inline pkg/nad/masterplugin.go:455]: also validate that it's not less than 0

File: pkg/nad/masterplugin.go [hunk 27]
```diff
@@ -393,3 +394,132 @@ func (plugin *MasterIPVlanPlugin) GetMasterPluginConfig() (*MasterPlugin, error)
...
+	}
+
+	plugin.masterPlugin.FailOverMac = failOverMac
+
+	return plugin
+}
+
+// WithMiimon defines Miimon configuration to MasterBondPlugin.
+func (plugin *MasterBondPlugin) WithMiimon(miimon int) *MasterBondPlugin {
+	if miimon <= 0 {
```
Comment (evgenLevin) [inline pkg/nad/masterplugin.go:468]: it could be 0

File: pkg/nad/masterplugin.go [hunk 28]
```diff
@@ -393,3 +394,132 @@ func (plugin *MasterIPVlanPlugin) GetMasterPluginConfig() (*MasterPlugin, error)
...
+		glog.V(100).Infof("error adding incorrect miimon value to MasterBondPlugin")
+
+		plugin.errorMsg = "error adding incorrect miimon value to MasterBondPlugin"
+	}
+
+	plugin.masterPlugin.Miimon = fmt.Sprintf("%d", miimon)
```
Comment (evgenLevin) [inline pkg/nad/masterplugin.go:474]: usually using  strconv.Itoa(int), but ok if it works

File: pkg/nad/nadtypes.go [hunk 29]
```diff
@@ -39,20 +40,27 @@ type (
 		Group            int               `json:"group,omitempty"`
 		MultiQueue       bool              `json:"multiQueue,omitempty"`
 		SelinuxContext   string            `json:"selinuxcontext,omitempty"`
+		Trust            bool              `json:"trust,omitempty"`
```
Comment (evgenLevin) [inline pkg/nad/nadtypes.go:43]: Why did you add "trust"?

File: pkg/nad/masterplugin.go [hunk 30]
```diff
@@ -393,3 +394,132 @@ func (plugin *MasterIPVlanPlugin) GetMasterPluginConfig() (*MasterPlugin, error)
...
+
+		builder.errorMsg = "NewMasterBondPlugin name is empty"
+	}
+
+	return &builder
+}
+
+// WithLinksInContainer defines linksInContainer configuration to MasterBondPlugin.
+func (plugin *MasterBondPlugin) WithLinksInContainer(linksInContainer bool) *MasterBondPlugin {
+	if reflect.TypeOf(linksInContainer).Kind() != reflect.Bool {
```
Comment (evgenLevin) [inline pkg/nad/masterplugin.go:442]: remove all validations here

**Comment by evgenLevin** [comment]
lgtm

File: pkg/nad/masterplugin.go [hunk 31]
```diff
@@ -393,3 +393,126 @@ func (plugin *MasterIPVlanPlugin) GetMasterPluginConfig() (*MasterPlugin, error)
...
+// NewMasterBondPlugin creates new instance of MasterBondPlugin.
+func NewMasterBondPlugin(name, mode string) *MasterBondPlugin {
+	validModes := map[string]bool{
+		"balance-rr":    true,
+		"active-backup": true,
+		"balance-xor":   true,
+	}
+
+	glog.V(100).Infof(
+		"Initializing new MasterBondPlugin structure %s", name)
```
Comment (kononovn) [inline pkg/nad/masterplugin.go:412]: Please add mode to the log:
glog.V(100).Infof("Initializing new MasterBondPlugin structure %s with mode %s", name, mode)

File: pkg/nad/masterplugin.go [hunk 32]
```diff
@@ -393,3 +393,126 @@ func (plugin *MasterIPVlanPlugin) GetMasterPluginConfig() (*MasterPlugin, error)
...
+
+		builder.errorMsg = "NewMasterBondPlugin name is empty"
+	}
+
+	return &builder
+}
+
+// WithLinksInContainer defines linksInContainer configuration to MasterBondPlugin.
+func (plugin *MasterBondPlugin) WithLinksInContainer(linksInContainer bool) *MasterBondPlugin {
+	plugin.masterPlugin.LinksInContainer = linksInContainer
```
Comment (kononovn) [inline pkg/nad/masterplugin.go:441]: Please add log:
glog.V(100).Infof("Adding LinksInContainer %d to MasterBoundPlugin %s", linksInContainer, plugin.masterPlugin.Name)

File: pkg/nad/masterplugin.go [hunk 33]
```diff
@@ -393,3 +393,126 @@ func (plugin *MasterIPVlanPlugin) GetMasterPluginConfig() (*MasterPlugin, error)
...
+
+	return plugin
+}
+
+// WithFailOverMac defines FailOverMac configuration to MasterBondPlugin.
+func (plugin *MasterBondPlugin) WithFailOverMac(failOverMac int) *MasterBondPlugin {
```
Comment (kononovn) [inline pkg/nad/masterplugin.go:447]: Please add log message

File: pkg/nad/masterplugin.go [hunk 34]
```diff
@@ -393,3 +393,126 @@ func (plugin *MasterIPVlanPlugin) GetMasterPluginConfig() (*MasterPlugin, error)
...
+		plugin.errorMsg = "error adding incorrect FailOverMac to MasterBondPlugin"
+	}
+
+	plugin.masterPlugin.FailOverMac = failOverMac
+
+	return plugin
+}
+
+// WithMiimon defines Miimon configuration to MasterBondPlugin.
+func (plugin *MasterBondPlugin) WithMiimon(miimon int) *MasterBondPlugin {
```
Comment (kononovn) [inline pkg/nad/masterplugin.go:460]: Please add log message

File: pkg/nad/masterplugin.go [hunk 35]
```diff
@@ -393,3 +393,126 @@ func (plugin *MasterIPVlanPlugin) GetMasterPluginConfig() (*MasterPlugin, error)
...
+		plugin.errorMsg = "error adding incorrect miimon value to MasterBondPlugin"
+	}
+
+	plugin.masterPlugin.Miimon = fmt.Sprintf("%d", miimon)
+
+	return plugin
+}
+
+// WithLinks defines Links configuration to MasterBondPlugin.
+func (plugin *MasterBondPlugin) WithLinks(links []Link) *MasterBondPlugin {
```
Comment (kononovn) [inline pkg/nad/masterplugin.go:473]: Please add log message

Comment (kononovn) [inline pkg/nad/masterplugin.go:473]: What is links is empty list. Please verify that links is not empty list

File: pkg/nad/masterplugin.go [hunk 36]
```diff
@@ -393,3 +393,126 @@ func (plugin *MasterIPVlanPlugin) GetMasterPluginConfig() (*MasterPlugin, error)
...
+		plugin.errorMsg = "error adding empty links to MasterBondPlugin"
+	}
+
+	plugin.masterPlugin.Links = links
+
+	return plugin
+}
+
+// WithCapabilities defines Capabilities configuration to MasterBondPlugin.
+func (plugin *MasterBondPlugin) WithCapabilities(capabilities *Capability) *MasterBondPlugin {
```
Comment (kononovn) [inline pkg/nad/masterplugin.go:486]: Please add log message.

File: pkg/nad/masterplugin.go [hunk 37]
```diff
@@ -393,3 +393,126 @@ func (plugin *MasterIPVlanPlugin) GetMasterPluginConfig() (*MasterPlugin, error)
...
+		plugin.errorMsg = "error adding empty capabilities to MasterBondPlugin"
+	}
+
+	plugin.masterPlugin.Capabilities = capabilities
+
+	return plugin
+}
+
+// WithIPAM defines IPAM configuration to MasterBondPlugin.
+func (plugin *MasterBondPlugin) WithIPAM(ipam *IPAM) *MasterBondPlugin {
```
Comment (kononovn) [inline pkg/nad/masterplugin.go:499]: Please add log message

**Review by kononovn — CHANGES_REQUESTED** [review]
Please fix few comments above.

File: pkg/nad/masterplugin.go [hunk 38]
```diff
@@ -277,7 +277,7 @@ func (plugin *MasterVlanPlugin) WithLinkInContainer() *MasterVlanPlugin {
 		return plugin
 	}
 
-	plugin.masterPlugin.LinkInContainer = true
+	plugin.masterPlugin.LinksInContainer = true
```
Comment (kononovn) [inline pkg/nad/masterplugin.go:280]: LinkInContainer and LinksInContainer are two different parameters. Please revert it back to it's original state

File: pkg/nad/masterplugin.go [hunk 39]
```diff
@@ -380,7 +380,7 @@ func (plugin *MasterIPVlanPlugin) WithLinkInContainer() *MasterIPVlanPlugin {
...
```
Comment (kononovn) [inline pkg/nad/masterplugin.go:383]: LinkInContainer and LinksInContainer are two different parameters. Please revert it back to it's original state

File: pkg/nad/nadtypes.go [hunk 40]
```diff
@@ -27,7 +28,7 @@ type (
 		Bridge           string            `json:"bridge,omitempty"`
 		Master           string            `json:"master,omitempty"`
 		Vlan             string            `json:"vlan,omitempty"`
-		Mtu              string            `json:"mtu,omitempty"`
+		Mtu              int               `json:"mtu,omitempty"`
```
Comment (kononovn) [inline pkg/nad/nadtypes.go:31]: Please revert it back to string

File: pkg/nad/nadtypes.go [hunk 41]
```diff
@@ -43,16 +44,22 @@ type (
...
-		Name            string    `json:"name,omitempty"`
-		Type            string    `json:"type,omitempty"`
-		Master          string    `json:"master,omitempty"`
-		Mode            string    `json:"mode,omitempty"`
-		Plugins         *[]Plugin `json:"plugins,omitempty"`
-		Bridge          string    `json:"bridge,omitempty"`
-		Ipam            *IPAM     `json:"ipam,omitempty"`
-		LinkInContainer bool      `json:"linkInContainer,omitempty"`
-		VlanID          uint16    `json:"vlanId,omitempty"`
+		CniVersion       string      `json:"cniVersion,omitempty"`
```
Comment (kononovn) [inline pkg/nad/nadtypes.go:47]: Please remove all unused parameters from this struct

**Review by kononovn — CHANGES_REQUESTED** [review]
Please fix few comments above

File: pkg/nad/masterplugin.go [hunk 42]
```diff
@@ -393,3 +393,127 @@ func (plugin *MasterIPVlanPlugin) GetMasterPluginConfig() (*MasterPlugin, error)
...
+		builder.errorMsg = "NewMasterBondPlugin name is empty"
+	}
+
+	return &builder
+}
+
+// WithLinksInContainer defines linksInContainer configuration to MasterBondPlugin.
+func (plugin *MasterBondPlugin) WithLinksInContainer(linksInContainer bool) *MasterBondPlugin {
+	plugin.masterPlugin.LinksInContainer = linksInContainer
+	glog.V(100).Infof("Adding LinksInContainer %d to MasterBoundPlugin %s", linksInContainer,
```
Comment (kononovn) [inline pkg/nad/masterplugin.go:441]: Please move this log up to the first line of the function.

File: pkg/nad/masterplugin.go (hunk 33 above)
Comment (kononovn) [reply pkg/nad/masterplugin.go:448]: I don't see any log here. Please add

File: pkg/nad/nadtypes.go (hunk 40 above)
Comment (kononovn) [reply pkg/nad/nadtypes.go:31]: @evgenLevin Please check

File: pkg/nad/masterplugin.go (hunk 34 above)
Comment (kononovn) [reply pkg/nad/masterplugin.go:461]: I don't see any log here. Please add

File: pkg/nad/masterplugin.go (hunk 35 above)
Comment (kononovn) [reply pkg/nad/masterplugin.go:474]: I don't see any log here. Please add log

File: pkg/nad/masterplugin.go (hunk 36 above)
Comment (kononovn) [reply pkg/nad/masterplugin.go:487]: I don't see any log here. Please add log.

File: pkg/nad/masterplugin.go (hunk 37 above)
Comment (kononovn) [reply pkg/nad/masterplugin.go:500]: I don't see any log here. Please add log.

File: pkg/nad/masterplugin.go [hunk 43]
```diff
@@ -393,3 +393,138 @@ func (plugin *MasterIPVlanPlugin) GetMasterPluginConfig() (*MasterPlugin, error)
...
+	}
+
+	plugin.masterPlugin.Ipam = ipam
+
+	return plugin
+}
+
+// GetMasterPluginConfig returns master plugin if error is not occur.
+func (plugin *MasterBondPlugin) GetMasterPluginConfig() (*MasterPlugin, error) {
+	glog.V(100).Infof("Adding GetMasterPluginConfig to MasterBoundPlugin")
```
Comment (kononovn) [inline pkg/nad/masterplugin.go:523]: Invalid log. Please remove.

File: pkg/nad/nadtypes.go (hunk 40 above)
Comment (gkopels) [reply pkg/nad/nadtypes.go:31]: changed back to string

**Review by kononovn — APPROVED** [review]
/lgtm
